### PR TITLE
Fix Selecting a map resets player color

### LINKED ISSUE #11898: Selecting a map with zero player spots resets player color in settings.

### PR DESCRIPTION
**Pull request by rob-v**
Closes #11898.

### CODE REVIEW

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 1]
```diff
@@ -712,7 +712,7 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
 						targetClient.Color = SanitizePlayerColor(server, newColor, targetClient.Index, conn);
 
 						// Only update player's preferred color if new color is valid
-						if (newColor == targetClient.Color)
+						if (newColor.RGB == targetClient.Color.RGB)
```
Comment (pchote) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: Please fix this inside HSLColor's `operator ==` so that everyone can benefit from the fix.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Game/Server/Server.cs [hunk 2]
```diff
@@ -306,7 +306,7 @@ void ValidateClient(Connection newConn, string data)
 					IpAddress = ((IPEndPoint)newConn.Socket.RemoteEndPoint).Address.ToString(),
 					Index = newConn.PlayerIndex,
 					Slot = LobbyInfo.FirstEmptySlot(),
-					PreferredColor = handshake.Client.Color,
+					PreferredColor = handshake.Client.PreferredColor,
```
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: I remember having a specific reason for using `Color` here, but I can't remember what it was.  Probably working around some even older bug that is no longer valid.

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs (hunk 1 above)
Comment (rob-v) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: Updated.

**Review by chrisforbes — APPROVED** [review]

File: OpenRA.Game/Graphics/HSLColor.cs [hunk 3]
```diff
@@ -95,7 +95,8 @@ public static bool TryParseRGB(string value, out Color color)
 
 		public static bool operator ==(HSLColor me, HSLColor other)
 		{
-			return me.H == other.H && me.S == other.S && me.L == other.L;
+			// binary floating point numbers (float, double) calculations can yield the same RGB color created by different functions with little different HSL representation
```
Comment (pchote) [inline OpenRA.Game/Graphics/HSLColor.cs]: Style nit: Comments should start with a capital letter.

**Review by pchote — APPROVED** [review]
:+1: with that minor fix.

Note that a much more obvious effect of this bug, which is what we should mention in the changelog for this, is that the server will reset your color when you change slots.

**Comment by rob-v** [comment]
Updated.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/50b7d3994f9024d17d867c3060f9dd8e74fc657c

**Pull request merged**